File: client/Assets/Plugins/Android/java/com/shiyuegame/httprequest/DownloadConfReader.java
package com.shiyuegame.httprequest;


import com.shiyuegame.util.SYUtil;

import java.io.BufferedReader;
import java.io.FileReader;
import java.util.Queue;
import java.util.concurrent.LinkedBlockingQueue;

public class DownloadConfReader {

    private Queue<VersionAsset> list = new LinkedBlockingQueue<VersionAsset>();

    public DownloadConfReader(String downloadConf) {
        Parse(downloadConf);
    }

    private void Parse(String path) {
        try {
            BufferedReader dis = new BufferedReader(new FileReader(path));
            String line = dis.readLine();
            while(line != null) {
                String [] items = line.split("<#>");
                if (items.length != 4)
                    continue;
                VersionAsset asset = new VersionAsset(items[0], items[1], items[2], items[3]);
                list.add(asset);
                line = dis.readLine();
            }
            dis.close();
            SYUtil.Log("==========list size:" + list.size());
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    public int GetSize() {
        synchronized (list) {
            return list.size();
        }
    }

    public VersionAsset SyncPoll() {
        synchronized (list) {
            if (list != null) {
                return list.poll();
            } else {
                return null;
            }
        }
    }

    public  void SyncAdd(VersionAsset asset) {
        synchronized (list) {
            if (list != null) {
                list.add(asset);
            }
        }
    }
    public  void SyncClear() {
        if (list == null) {
            return;
        }
        synchronized (list) {
            list.clear();
            list = new LinkedBlockingQueue<VersionAsset>();
        }
    }
}

File: client/Assets/Plugins/iOS/SySwiftNetWork/DownloadConfReader.swift
//
//  DownloadConfReader.swift
//  TestSwift
//
//  Created by huangyuqiu on 2023/3/15.
//

import Foundation

class DownloadConfReader {
    
    private var list:Queue<VersionAsset>
    private let listLock = NSLock()

    init(path:String) {
        list = Queue<VersionAsset>()
        self.Parse(path:path)
    }
    
    private func Parse(path: String) -> Void {
        do {
            let fileUrl = URL(fileURLWithPath: path)
            let content = try String(contentsOf: fileUrl, encoding: .utf8)
            let lines = content.split(separator: "\n")
            for line in lines {
//                print("line:\(line)")
                let items = line.trimmingCharacters(in: .whitespaces).components(separatedBy: "<#>")
                if items.count != 4 {
                    continue
                }
                let asset = VersionAsset(path:items[0], size:items[2], patchVersion:items[1], md5:items[3])
                list.enqueue(asset)
                // print("/\(asset.patchVersion)/\(asset.path) \(asset.size) \(asset.md5)")
            }
            print("ios conf list size:\(list.count)")
        } catch _ {
            print("Download Config Parse Error")
        }
    }
    
    public func GetSize() -> Int {
        var count = 0
        listLock.lock()
        count = list.count
        listLock.unlock()
        return count
    }
    
    public func SyncPoll() -> VersionAsset? {
        var asset:VersionAsset?
        listLock.lock()
        if !list.isEmpty {
            asset = list.dequeue()
        }
        listLock.unlock()
        return asset
    }
    
    public func SyncAdd(asset:VersionAsset) -> Void {
        listLock.lock()
        list.enqueue(asset)
        listLock.unlock()
    }
    
    public func SyncClear() -> Void {
        if list.isEmpty {
            return
        }
        listLock.lock()
        list.clear()
        list = Queue<VersionAsset>()
        listLock.unlock()
    }
}

class VersionAsset {
    private(set) var path: String
    private(set) var size: String
    private(set) var patchVersion: String
    private(set) var md5: String
    
    init(path: String, size: String, patchVersion: String, md5: String) {
        self.path = path
        self.size = size
        self.patchVersion = patchVersion
        self.md5 = md5
    }
}

public struct Queue<T> {
    fileprivate var array = [T]()
    public var isEmpty: Bool {
        return array.isEmpty
    }
    public var count: Int {
        return array.count
    }
    public mutating func enqueue(_ element: T) {
        array.append(element)
    }
    public mutating func dequeue() -> T? {
        if isEmpty {
           return nil
        } else {
           return array.removeFirst()
        }
     }
    public mutating func clear() -> Void {
        array.removeAll()
    }
        
    public var front: T? {
        return array.first
    }
}
